### PR TITLE
Added verifyBeforeUpdateEmail method to Web, iOS, Android

### PR DESCRIPTION
**Pull request by ewwwgiddings**
Pull request to handle this feature request:
https://github.com/capawesome-team/capacitor-firebase/issues/739

## Pull request checklist

Please check if your PR fulfills the following requirements:

- [ ] The changes have been tested successfully.
- [ ] A changeset has been created (`npm run changeset`).
- [ ] I have read and followed the [pull request guidelines](https://capawesome.io/contributing/pull-requests/).

### CODE REVIEW

File: packages/authentication/android/src/main/java/io/capawesome/capacitorjs/plugins/firebase/authentication/FirebaseAuthentication.java [hunk 1]
```diff
@@ -587,6 +587,16 @@ public void updateEmail(FirebaseUser user, @NonNull String newEmail, @NonNull Ru
             );
     }
 
+    public void verifyBeforeUpdateEmail(FirebaseUser user, @NonNull String newEmail, @NonNull ActionCodeSettings actionCodeSettings, @NonNull Runnable callback) {
```
Comment (robingenz) [inline packages/authentication/android/src/main/java/io/capawesome/capacitorjs/plugins/firebase/authentication/FirebaseAuthentication.java:590]:
```suggestion
    public void verifyBeforeUpdateEmail(FirebaseUser user, @NonNull String newEmail, @NonNull ActionCodeSettings actionCodeSettings, @NonNull EmptyResultCallback callback) {
```

File: packages/authentication/android/src/main/java/io/capawesome/capacitorjs/plugins/firebase/authentication/FirebaseAuthentication.java [hunk 2]
```diff
@@ -587,6 +587,16 @@ public void updateEmail(FirebaseUser user, @NonNull String newEmail, @NonNull Ru
...
+        user
+            .verifyBeforeUpdateEmail(newEmail, actionCodeSettings)
+            .addOnCompleteListener(
+                task -> {
+                    callback.run();
+                }
+            );
```
Comment (robingenz) [inline packages/authentication/android/src/main/java/io/capawesome/capacitorjs/plugins/firebase/authentication/FirebaseAuthentication.java:597]:
```suggestion
            .addOnSuccessListener(unused -> callback.success())
            .addOnFailureListener(exception -> callback.error(exception));
```

I guess this method can throw an excpetion or not? If so, then we should handle this.

**Review by robingenz — CHANGES_REQUESTED** [review]
This already looks very good. Please take a look at my comments.

**Comment by ewwwgiddings** [comment]
Thanks! Wasn't sure how far I was off with those.

I just implemented all your suggestions now and hit the re-request review button.

**Comment by ewwwgiddings** [comment]
I also just updated the `newEmail` definition to what you liked and added an example in the README.

File: packages/authentication/android/src/main/java/io/capawesome/capacitorjs/plugins/firebase/authentication/FirebaseAuthentication.java (hunk 2 above)
Comment (robingenz) [reply packages/authentication/android/src/main/java/io/capawesome/capacitorjs/plugins/firebase/authentication/FirebaseAuthentication.java]: Thanks for pointing that out. Feel free to create a new PR for the other methods. For now, just push that single change.

File: packages/authentication/android/src/main/java/io/capawesome/capacitorjs/plugins/firebase/authentication/FirebaseAuthenticationPlugin.java [hunk 3]
```diff
@@ -840,6 +840,36 @@ public void updateEmail(PluginCall call) {
...
+            }
+
+            JSObject settings = call.getObject("actionCodeSettings");
+            if (settings == null) {
+                call.reject(ERROR_ACTION_CODE_SETTINGS_MISSING);
+                return;
+            }
+            ActionCodeSettings actionCodeSettings = FirebaseAuthenticationHelper.createActionCodeSettings(settings);
+            
+            implementation.verifyBeforeUpdateEmail(user, newEmail, actionCodeSettings, () -> call.resolve());
```
Comment (robingenz) [inline packages/authentication/android/src/main/java/io/capawesome/capacitorjs/plugins/firebase/authentication/FirebaseAuthenticationPlugin.java:865]: You should also deal with the errors here (for example see `confirmVerificationCode`).

**Review by robingenz — CHANGES_REQUESTED** [review]

**Comment by ewwwgiddings** [comment]
Just pushed a commit that adds the callback like the example method.

**Comment by ewwwgiddings** [comment]
Just fixed the prettier issues as well. There were around 25 in unrelated files when using `--write` on the prettier command as a side note for some time in the future maybe.

File: packages/authentication/android/src/main/java/io/capawesome/capacitorjs/plugins/firebase/authentication/FirebaseAuthenticationPlugin.java [hunk 4]
```diff
@@ -840,6 +840,50 @@ public void updateEmail(PluginCall call) {
...
+            }
+
+            JSObject settings = call.getObject("actionCodeSettings");
+            if (settings == null) {
+                call.reject(ERROR_ACTION_CODE_SETTINGS_MISSING);
+                return;
+            }
+            ActionCodeSettings actionCodeSettings = FirebaseAuthenticationHelper.createActionCodeSettings(settings);
+
+            NonEmptyResultCallback callback = new NonEmptyResultCallback() {
```
Comment (robingenz) [inline packages/authentication/android/src/main/java/io/capawesome/capacitorjs/plugins/firebase/authentication/FirebaseAuthenticationPlugin.java:865]:
```suggestion
            EmptyResultCallback callback = new EmptyResultCallback() {
```

File: packages/authentication/android/src/main/java/io/capawesome/capacitorjs/plugins/firebase/authentication/FirebaseAuthenticationPlugin.java [hunk 5]
```diff
@@ -840,6 +840,50 @@ public void updateEmail(PluginCall call) {
...
+                @Override
+                public void success(Result result) {
+                    call.resolve(result.toJSObject());
+                }
```
Comment (robingenz) [inline packages/authentication/android/src/main/java/io/capawesome/capacitorjs/plugins/firebase/authentication/FirebaseAuthenticationPlugin.java:869]:
```suggestion
                public void success() {
                    call.resolve();
                }
```

**Review by robingenz — CHANGES_REQUESTED** [review]
These should be the last changes.

**Comment by egiddings-dealeron** [comment]
Changes pushed

**Comment by robingenz** [comment]
@ewwwgiddings Great, thank you! Two points remain before we can merge:
1. Create a changeset with the command`npm run changeset`. Please use the following text: ``feat: add `verifyBeforeUpdateEmail` method``
2. Test your implementation using this dev release:
	```
	npm i https://pkg.pr.new/capawesome-team/capacitor-firebase/@capacitor-firebase/authentication@741
	```

**Comment by ewwwgiddings** [comment]
Pushed the changeset. As far as testing I'm currently unable to test anything but web with my setup. Eventually I'll get to a point I have all the environments set up but atm it's probably a couple months away.

File: packages/authentication/src/definitions.ts [hunk 6]
```diff
@@ -780,6 +788,24 @@ export interface UpdateEmailOptions {
...
+   *
+   * @since 6.3.0
+   */
+  newEmail: string;
+  /**
+   * The action code settings
+   *
+   * @since 6.3.0
+   */
+  actionCodeSettings?: ActionCodeSettings;
```
Comment (robingenz) [inline packages/authentication/src/definitions.ts:806]:
```suggestion
  actionCodeSettings: ActionCodeSettings;
```

**Review by robingenz — APPROVED** [review]

**Comment by robingenz** [comment]
@ewwwgiddings No problem. I will test it on Android and iOS before the next release. Thank you for your contribution!

**Pull request merged**